### PR TITLE
Game dev - game system

### PR DESCRIPTION
**Pull request by bamboolord**

### CODE REVIEW

File: game/workers/asset-loader.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ export default function loadAssets() {
       });
 
     // Map
-    const load3 = loadAsset(['Images/template1.png', 'Images/template1_collision.png'])
+    const load3 = loadAsset(['Images/template1_collision.png', 'Images/template1_collision.png'])
```
Comment (sharpstorm) [inline game/workers/asset-loader.js:38]: File Wrong

File: game/ui/ui-game.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import { createElement } from './ui-utils.js';
+import UIElement, { UIAnchor } from './ui-element.js';
+import GameManager from '../managers/game-manager.js';
```
Comment (sharpstorm) [inline game/ui/ui-game.js:3]: Should change to attach click listener instead. Game Manager only used once, can further separate elements

File: game/net/server/game-data-sender.js [hunk 3]
```diff
@@ -89,6 +99,7 @@ const handlers = {
   'chat-echo': buildChatEcho,
   'checkers': buildCheckers,
   'voice-channel-data': buildVoiceChannelData,
+  'gameLobby-echo': buildGameLobbyUpdate,
```
Comment (sharpstorm) [inline game/net/server/game-data-sender.js:102]: Need to normalise OpCode

File: game/net/server/game-data-handler.js [hunk 4]
```diff
@@ -83,6 +195,7 @@ const handlers = {
   'checkers': handleCheckersGame,
   'join-voice': handleJoinVoice,
   'disconnect-voice': handleDisconnectVoice,
+  'gameLobby': handleGameLobby,
```
Comment (sharpstorm) [inline game/net/server/game-data-handler.js:198]: Normalise OpCode

File: game/net/server/game-data-handler.js [hunk 5]
```diff
@@ -75,6 +75,118 @@ function handleDisconnectVoice(data, conn) {
   conn.send(packet);
 }
 
+// eslint-disable-next-line no-unused-vars
```
Comment (sharpstorm) [inline game/net/server/game-data-handler.js:78]: Need to change how packet handling is done here - Too many subcommands under a single opcode

File: game/net/server/game-data-handler.js [hunk 6]
```diff
@@ -58,7 +58,7 @@ function handleChat(data, conn) {
 
 function handleCheckersGame(data, conn) {
   console.log(data);
-  checkersMove(data);
+  CheckersGame.getInstance().checkersMove(data);
```
Comment (sharpstorm) [inline game/net/server/game-data-handler.js:61]: Need to generalise

File: game/net/client/game-data-sender.js [hunk 7]
```diff
@@ -50,6 +61,7 @@ const handlers = {
   'checkers': buildCheckersUpdate,
   'join-voice': buildEmptyPacket,
   'disconnect-voice': buildEmptyPacket,
+  'gameLobby': buildGameLobbyUpdate,
```
Comment (sharpstorm) [inline game/net/client/game-data-sender.js:64]: Normalise OpCode

File: game/net/client/game-data-handler.js [hunk 8]
```diff
@@ -82,6 +122,7 @@ const handlers = {
   'chat-echo': handleChatEcho,
   'checkers': handleCheckersEcho,
   'voice-channel-data': handleVoiceChannelData,
+  'gameLobby-echo': handleGameLobby,
```
Comment (sharpstorm) [inline game/net/client/game-data-handler.js:125]: Normalise OpCode

File: game/net/client/game-data-handler.js [hunk 9]
```diff
@@ -64,14 +65,53 @@ function handleChatEcho(data, conn) {
...
-  checkersMove(data);
+  CheckersGame.getInstance().checkersMove(data);
 }
 
 function handleVoiceChannelData(data, conn) {
   const voiceUsers = data.users;
   GameManager.getInstance().getVoiceChannelManager().updateChannelUsers(voiceUsers);
 }
 
+function handleGameLobby(data, conn) {
```
Comment (sharpstorm) [inline game/net/client/game-data-handler.js:76]: Very Messy need to tidy up

File: game/net/client/game-data-handler.js [hunk 10]
```diff
@@ -64,14 +65,53 @@ function handleChatEcho(data, conn) {
 
 function handleCheckersEcho(data, conn) {
   console.log(data);
-  checkersMove(data);
+  CheckersGame.getInstance().checkersMove(data);
```
Comment (sharpstorm) [inline game/net/client/game-data-handler.js:68]: Need to Generalise

File: game/models/map.js [hunk 11]
```diff
@@ -35,6 +36,21 @@ export default class Map {
     }
   }
 
+  getFuniture(x, y) {
```
Comment (sharpstorm) [inline game/models/map.js:39]: Spelling

Comment (sharpstorm) [reply game/models/map.js:39]: Why not do this during initialization where collision is processed

File: game/models/map.js [hunk 12]
```diff
@@ -10,6 +10,7 @@ export default class Map {
     this.heightGrids = heightGrids;
 
     this.initCollisionMap();
+    document.addEventListener('click', this.getColor);
```
Comment (sharpstorm) [inline game/models/map.js:13]: No Modifying DOM in Models

File: game/managers/world-manager.js [hunk 13]
```diff
@@ -4,6 +4,119 @@ export default class WorldManager {
   constructor() {
     this.peerIdToUidMap = {};
     this.voiceChannelUsers = new Set();
+    this.gameLobbies = {};
```
Comment (sharpstorm) [inline game/managers/world-manager.js:7]: Lobby system is coded to checkers. Should be the other way round

File: game/managers/game-manager.js [hunk 14]
```diff
@@ -166,10 +171,275 @@ class TextChannelManager {
...
 }
 
+class BoardGameManager {
+  constructor() {
+    this.gameList = [];
+    this.uiLayer = undefined; // cant define here. cyclic
+    this.currentGame = undefined; // USELESS? NOPE
+    this.gameState = undefined; // change to number
+    this.tableID = undefined;
+    /* GAME STATES
```
Comment (sharpstorm) [inline game/managers/game-manager.js:181]: Make this list into enum

File: game/managers/game-manager.js [hunk 15]
```diff
@@ -166,10 +171,275 @@ class TextChannelManager {
...
+          } else {
+            this.displayPage(0);
+          }
+          break;
+        default:
+      }
+    }
+  }
+
+  getGame(gameName) {
```
Comment (sharpstorm) [inline game/managers/game-manager.js:240]: Use Lambda Find

File: game/managers/game-manager.js [hunk 16]
```diff
@@ -166,10 +171,275 @@ class TextChannelManager {
...
+    let i;
+    for (i = 0; i < this.gameList.length; i += 1) {
+      if (this.gameList[i].gameName === gameName) {
+        return this.gameList[i];
+      }
+    }
+    return undefined;
+  }
+
```
Comment (sharpstorm) [inline game/managers/game-manager.js:249]: Should change ConstructorName, not advisable to use

File: game/managers/animation-manager.js [hunk 17]
```diff
@@ -89,6 +89,27 @@ class UILayer {
   }
 }
 
+class GameLayer {
+  constructor() {
+    this.drawables = [];
+  }
+
+  register(game) {
```
Comment (sharpstorm) [inline game/managers/animation-manager.js:97]: Design: Game layer drawables should be independent of GameManager --> Possible to remove GameManager import

**Review by sharpstorm — CHANGES_REQUESTED** [review]
Many problems

File: game/managers/world-manager.js (hunk 13 above)
Comment (sharpstorm) [reply game/managers/world-manager.js:7]: Generalised Lobby System

File: game/net/server/game-data-sender.js (hunk 3 above)
Comment (bamboolord) [reply game/net/server/game-data-sender.js:102]: fixed

File: game/net/server/game-data-handler.js (hunk 4 above)
Comment (bamboolord) [reply game/net/server/game-data-handler.js:198]: fixed

**Review by sharpstorm — APPROVED** [review]
Issues Resolved

**Pull request merged**